include/gridtools/thread_pool/concept.hpp

```diff
@@ -0,0 +1,70 @@
+/*
+ * GridTools
+ *
+ * Copyright (c) 2014-2019, ETH Zurich
+ * All rights reserved.
+ *
+ * Please, refer to the LICENSE file in the root directory.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+#pragma once
+
+/*
+ * The Concept of Thread Pool
+ *
+ * To be a thread pool a type should have the following functions to be available via ADL:
+ *   thread_pool_get_thread_num(pool);
+ *   thread_pool_get_max_threads(pool);
+ *   thread_pool_parallel_for_loop(pool, func, limit);
+ *
+ *   where `pool` is an instance of the type.
+ *
+ *   There maybe additional overloads of `thread_pool_parallel_for_loop` avaliable:
+ *     thread_pool_parallel_for_loop(pool, func, lim0, lim1);
+ *     thread_pool_parallel_for_loop(pool, func, lim0, lim1, lim2);
+ *     etc.
+ *   They are optional and could be provided for performance reasons.
+ */
+
+#include <tuple>
+
+#include "../common/stride_util.hpp"
+#include "../common/tuple_util.hpp"
+namespace gridtools {
+    namespace thread_pool {
+        namespace concept_impl_ {
+            template <class T>
+            auto get_thread_num(T const &obj) -> decltype(thread_pool_get_thread_num(obj)) {
+                return thread_pool_get_thread_num(obj);
+            }
+
+            template <class T>
+            auto get_max_threads(T const &obj) -> decltype(thread_pool_get_max_threads(obj)) {
+                return thread_pool_get_max_threads(obj);
+            }
+
+            template <class T, class F, class... Dims>
+            auto thread_pool_parallel_for_loop(T const &obj, F const &f, Dims... limits)
+                -> decltype(thread_pool_parallel_for_loop(obj, std::declval<void (*)(size_t)>, 0)) {
+                std::tuple<Dims...> lims{limits...};
+                thread_pool_parallel_for_loop(obj,
+                    [&, strides = stride_util::make_strides_from_sizes(lims)](auto index) {
```

Comment:
Am I right that this actually generates a loop nesting that is inverse of the OpenMP one? That is, first limit has stride one instead of last one? If yes, this should probably be fixed, as the loop nesting order can have very significant influence on the performance (for example if a single thread works on a contiguous chunk).

Reply:
Yes. You are right here. I documented the order of parameters now. So that the first is the innermost. However I was not aware that OpenMP spec applies any requirements to the implementation on execution order. My naive understanding was that if we have `#pragma parallel for` the order is undefined.

Reply:
It actually does. If you use parallel for with static scheduling (which seems to be the default in available implementations, but I actually just checked in the standard that it seems to be implementation defined), the whole linearized (in the case of collapse) iteration space is divided into contiguous chunks of about equal size and each thread works on such a chunk.